### ISSUE TITLE
Cross-backend fan-out declares logical schema over a physical stream (column leak; fatal under StreamCache FFI)

### ISSUE THREAD

**Issue by mesejo** [xorq-labs/xorq#2055]
## Summary

In a cross-backend fan-out (`into_backend` / `RemoteTable`), `register_and_transform_remote_tables` wraps the remote expression's stream with the expression's **logical** schema via `pa.RecordBatchReader.from_batches(<logical schema>, batches)`. If the physical stream ever carries a column the logical schema does not list, the declared schema and the batches disagree.

- On `main` (SafeTee path) this is silently masked: `read_record_batches` → `_select_and_cast` drops the extra column and emits `UserWarning: batch has extra columns not in schema, dropping: ['row_number']`.
- On `spike/add-batchcorder` (batchcorder `StreamCache`) the cache exports the declared (logical) schema over the Arrow C Stream FFI, then ingests a batch with an extra child, and **aborts the process** (`arrow-array ffi.rs: assertion failed: fields.len() == self.array.num_children()`, SIGABRT / exit 134).

So a benign-on-`main` schema/stream divergence becomes a fatal, uncatchable crash once the stream crosses an FFI boundary that trusts the declared schema.

## Reproduction

xorq `main` (`git+https://github.com/xorq-labs/xorq@f46446bd`), `duckdb==1.3.2`, `xorq-datafusion==0.2.7`, Python 3.13:

```bash
xorq init --path sklearn_template --template sklearn
xorq uv build sklearn_template/expr.py --builds-dir dist/template-builds
xorq uv run dist/template-builds/<hash> --format parquet
```

Output (exit 0 on `main`, warning fires):

```
xorq/backends/xorq_datafusion/__init__.py:794: UserWarning: batch has extra columns not in schema, dropping: ['row_number']
  schema, (_select_and_cast(batch, schema) for batch in batches)
```

The leaked column is `row_number`, an internal helper `train_test_splits` adds and the template logically `.drop`s.

### Important: this is NOT a `.drop` / SQL-generation bug

Earlier I suspected duckdb `1.3.2` retained `row_number` despite the drop. **That is wrong — verified:**

- xorq compiles `.drop("row_number")` to duckdb-native `SELECT "t2".* EXCLUDE ("row_number") FROM (...)`. The logical schema correctly omits `row_number`. The compiled SQL survives the ibis-yaml roundtrip unchanged.
- Raw duckdb honors `EXCLUDE` in the Arrow stream on both versions:
  ```
  duckdb 1.3.2:  relation=['a']  reader=['a']  batch=['a']
  duckdb 1.5.2:  relation=['a']  reader=['a']  batch=['a']
  ```
- Draining the actual leaking remote node — `RemoteTable.remote_expr.to_pyarrow_batches()` — on duckdb `1.3.2` yields a clean 65-column stream, no `row_number`.

### Where the divergence actually appears

The warning reproduces **only** through the isolated packaged subprocess (`xorq uv run` → `uv tool run --isolated --with <wheel> --with-requirements requirements.txt xorq run <build>`). Every in-process path on identical code + deps is **clean**:

- `expr.execute()`
- `expr.to_pyarrow_batches()` drain
- direct `xorq run <build> --format parquet` (both `main`/SafeTee and `spike`/StreamCache)

So the trigger is something in the isolated packaged-run environment / dep closure, not SQL generation and not the duckdb drop. Root cause of *why* `row_number` reaches the fan-out only under the isolated run is still open.

## Root cause (of the fragility, confirmed)

`python/xorq/expr/relations.py`, `register_and_transform_remote_tables`:

```python
ex      = arg.remote_expr
batches = ex.to_pyarrow_batches()                 # physical stream
schema  = ex.as_table().schema().to_pyarrow()     # logical schema
replicas = SafeTee.tee(batches, count)
...
reader  = pa.RecordBatchReader.from_batches(schema, batchess.pop())  # declares logical over physical
```

`from_batches` re-declares the logical schema over a stream whose batches may differ. The two are assumed equal but are not guaranteed to be. Same pattern in the Flight paths (`FlightExpr.to_rbr`, `FlightUDXF`), all `from_batches(<logical schema>, gen)`. Whatever produces a divergent batch (still under investigation for this case), the fan-out has no defense and propagates a reader that lies about its stream — harmless under SafeTee + `_select_and_cast`, fatal under StreamCache FFI.

## Proposed fix

Stop declaring a schema that may not match the batches.

1. **Producer-side coercion (preferred)** — project/cast each batch to the logical schema before wrapping, mirroring `_select_and_cast` at the source so declared schema and batches genuinely agree:

   ```python
   def _conform(batches, schema):
       names = schema.names
       for b in batches:
           if set(b.schema.names) - set(names):
               b = b.select(names)
           yield b.cast(schema)
   reader = pa.RecordBatchReader.from_batches(schema, _conform(batchess.pop(), schema))
   ```

2. **Use the reader's own schema** — `pa.RecordBatchReader.from_batches(batches.schema, ...)` (or pass the reader through unchanged). The declared schema then always matches the batches; coercion stays the consumer's job. This is the minimal fix already applied on `spike/add-batchcorder` for the `StreamCache` path.

Either removes the FFI abort and the silent column leak. Option 1 also keeps the surviving stream matching the declared/expected schema — the more correct contract.

## Environment
- xorq `main` (`f46446bd`)
- `duckdb==1.3.2` (also checked `1.5.2`), `xorq-datafusion==0.2.7`, `pyarrow==21.0.0`
- Python 3.13

**Comment by mesejo** [comment]
Updated: corrected the root-cause analysis. The `.drop` and its SQL are **not** the bug — xorq emits duckdb-native `SELECT * EXCLUDE("row_number")`, the logical schema correctly omits it, and duckdb 1.3.2 honors EXCLUDE in the Arrow stream (verified raw + via the actual remote node's `to_pyarrow_batches()`, both clean). The `row_number` warning reproduces **only** through the isolated packaged subprocess (`xorq uv run`); every in-process path on identical code+deps is clean. Why the column reaches the fan-out only under the isolated run is still open.
